### PR TITLE
network: wifi ssid support with nl80211

### LINKED ISSUE #8157: [RFE] IWD support for systemd-networkd
### Submission type

 - Request for enhancement (RFE)

## Description

IWD, the new Intel Wireless daemon got its first release now[1].
It has full dbus support and other neat features. 
I would like to suggest IWD support in `systemd-networkd`.

[1] https://www.kernel.org/pub/linux/network/wireless/

### PR DESCRIPTION
**Pull request by yuwata**
Closes #8157 and #9466.
Replaces #12842.

### CODE REVIEW

**Comment by victorenator** [comment]
It would be nice to add a WirelessType match option (NL80211_ATTR_IFTYPE) with values:
- ad-hoc (NL80211_IFTYPE_ADHOC)
- ap (NL80211_IFTYPE_AP)
- station (NL80211_IFTYPE_STATION)
- …

And add the default network files:

* 80-wlan-station.network
```ini
[Match]
WirelessType=station

[Network]
DHCP=yes
```

* 80-wlan-ap.network
```ini
[Match]
WirelessType=ap

[Network]
Address=0.0.0.0/24
DHCPServer=yes
IPMasquerade=yes
```

* 80-wlan-adhoc.network
```ini
[Match]
WirelessType=ad-hoc

[Network]
LinkLocalAddressing=yes
```

**Comment by yuwata** [comment]
@keszybz and @ssahani Could you review this? This blocks #13452.

**Comment by genofire** [comment]
nl80211 supports also WEP -> a nice to have till we good somethink like WPA(2) into nl80211

File: man/systemd.network.xml [hunk 1]
```diff
@@ -153,6 +153,24 @@
             </para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>SSID=</varname></term>
+          <listitem>
+            <para>A whitespace-separated list of shell-style globs matching the SSID of the currently
+            connected wireless network. If the list is prefixed with a "!", the test is inverted.
```
Comment (ssahani) [inline man/systemd.network.xml]: Not anything wrong but do we have to explicitly mention the '!' thing ?

File: src/libsystemd/sd-netlink/generic-netlink.c [hunk 2]
```diff
@@ -23,12 +28,12 @@ int sd_genl_socket_open(sd_netlink **ret) {
 static int lookup_id(sd_netlink *nl, sd_genl_family family, uint16_t *id);
 
 static int genl_message_new(sd_netlink *nl, sd_genl_family family, uint16_t nlmsg_type, uint8_t cmd, sd_netlink_message **ret) {
-        int r;
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
         struct genlmsghdr *genl;
         const NLType *genl_cmd_type, *nl_type;
         const NLTypeSystem *type_system;
         size_t size;
-        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
```
Comment (ssahani) [inline src/libsystemd/sd-netlink/generic-netlink.c:31]: Not strict we can put this in to reverse tree order to look good.

File: man/systemd.network.xml (hunk 1 above)
Comment (yuwata) [reply man/systemd.network.xml]: OK, but I'd like to leave that in another PR, as we should update all other entries in [Match] section.

**Comment by yuwata** [comment]
@ssahani Thank you for the review.

File: src/libsystemd/sd-netlink/generic-netlink.c (hunk 2 above)
Comment (yuwata) [reply src/libsystemd/sd-netlink/generic-netlink.c:31]: Fixed.

**Comment by yuwata** [comment]
@victorenator Why `Address=0.0.0.0/24` for the AP mode?

File: man/systemd.network.xml [hunk 3]
```diff
@@ -153,6 +153,14 @@
...
```
Comment (Pesa) [inline man/systemd.network.xml]: wouldn't "Wi-Fi" be more correct than "wireless"? the latter sounds more generic to me, and may include things like bluetooth, cellular, etc.

File: man/systemd.network.xml [hunk 4]
```diff
@@ -161,6 +161,16 @@
             </para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>BSSID=</varname></term>
+          <listitem>
+            <para>A whitespace-separated list of hardware address of the currently connected wireless
```
Comment (Pesa) [inline man/systemd.network.xml:167]: ditto

File: man/systemd.network.xml (hunk 3 above)
Comment (flo-wer) [reply man/systemd.network.xml]: Quoting Wikipedia: 
> Wi‑Fi is a trademark of the Wi-Fi Alliance, [...]

One can't use the term "Wi-Fi" without violating their trademark as the linux kernel does not implement anything certified but only a subsystem that controls IEEE 802.11 devices. It is bad enough that the widespread marketing creeps this term into code and technical documents (e.g. the ProxyARPWiFi= config that is named after the kernel setting).

The linux kernel rightfully calls this cfg80211 subsystem and the correct terminology is either Wireless LAN or WLAN.

Comment (Pesa) [reply man/systemd.network.xml]: Fair enough

**Comment by yuwata** [comment]
@keszybz Ping?

File: src/libsystemd/sd-netlink/generic-netlink.c [hunk 5]
```diff
@@ -158,3 +158,19 @@ int nlmsg_type_to_genl_family(sd_netlink *nl, uint16_t type, sd_genl_family *ret
 
         return 0;
 }
+
+int sd_genl_message_get_family(sd_netlink *nl, sd_netlink_message *m, sd_genl_family *family) {
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/generic-netlink.c:163]: Shoudl this be `const sd_netlink *, const sd_netlink_message *` ?

File: src/libsystemd/sd-netlink/netlink-message.c [hunk 6]
```diff
@@ -618,6 +618,32 @@ int sd_netlink_message_read(sd_netlink_message *m, unsigned short type, size_t s
         return r;
 }
 
+int sd_netlink_message_read_string_strdup(sd_netlink_message *m, unsigned short type, char **data) {
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/netlink-message.c:618]: `const sd_netlink_message *` ?

**Comment by yuwata** [comment]
@keszybz Thank you for the review.

**Pull request merged**